### PR TITLE
8255374: Add a dropReturn MethodHandle combinator

### PR DESCRIPTION
**Pull request by JornVernee**
Hi,

This patch adds a `dropReturn` combinator to `MethodHandles` that can be used to create a new method handle that drops the return value, from a given method handle. Similar to the following code:

```
MethodHandle target = ...;
Class<?> targetReturnType = target.type().returnType();
if (targetReturnType != void.class)
    target = filterReturnValue(target, empty(methodType(void.class, targetReturnType))); 
// use target
```

But as a short-hand.

Thanks,
Jorn

CSR link: https://bugs.openjdk.java.net/browse/JDK-8255398

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Testing

|     | Linux x32 | Linux x64 | Windows x64 | macOS x64 |
| --- | ----- | ----- | ----- | ----- |
| Build | ✔️ (1/1 passed) | ✔️ (5/5 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) |
| Test (tier1) |    |  ✔️ (9/9 passed) | ✔️ (9/9 passed) | ✔️ (9/9 passed) |

### Issue
 * [JDK-8255374](https://bugs.openjdk.java.net/browse/JDK-8255374): Add a dropReturn MethodHandle combinator

### Reviewers
 * [Claes Redestad](https://openjdk.java.net/census#redestad) (@cl4es - **Reviewer**) ⚠️ Review applies to 22ae65d0e6b96b980a09b5ad4475db4a05077ec0

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/866/head:pull/866`
`$ git checkout pull/866`

### CODE REVIEW

File: src/java.base/share/classes/java/lang/invoke/MethodHandles.java [hunk 1]
```diff
@@ -5204,6 +5205,29 @@ public static MethodHandle dropArgumentsToMatch(MethodHandle target, int skip, L
...
+     */
+    public static MethodHandle dropReturn(MethodHandle target) {
+        Objects.requireNonNull(target);
+        MethodType oldType = target.type();
+        Class<?> oldReturnType = oldType.returnType();
+        if (oldReturnType == void.class)
+            return target;
+        MethodType newType = oldType.changeReturnType(void.class);
+        BoundMethodHandle result = target.rebind();
+        LambdaForm lform = result.form;
```
Comment (cl4es) [inline src/java.base/share/classes/java/lang/invoke/MethodHandles.java:5225]:
```suggestion
        LambdaForm lform = result.editor().filterReturnForm(V_TYPE, true);
```

File: src/java.base/share/classes/java/lang/invoke/MethodHandles.java [hunk 2]
```diff
@@ -5204,6 +5205,29 @@ public static MethodHandle dropArgumentsToMatch(MethodHandle target, int skip, L
...
+        lform = lform.editor().filterReturnForm(V_TYPE, true);
```
Comment (cl4es) [inline src/java.base/share/classes/java/lang/invoke/MethodHandles.java:5226]:
```suggestion
```

**Review by cl4es — APPROVED** [review]
LGTM, pending CSR.

A minor simplification suggested inline.

**Comment by rose00** [comment]
I don't mind shorthands, but the existing idiom is shorter than advertised,
a one-liner assuming the MH is already bound to a var:

```
target = target.asType(target.type().changeReturnType(void.class));
```

The API has good short circuits already; no new objects are
created if the MH is already void-returning.

Perhaps this RFE could be adjusted to `MethodHandle::changeReturnType`?
Then it could be used for dropping returns *or* casting them to other types.

**Comment by JornVernee** [comment]
@rose00 My bad. You're right, I realized we could also use `asType` later on (I did call it out in the CSR), but It's still a bit more wordy than what I'm proposing.

`dropReturn` seemed like a good choice since we already have `dropArguments`. WRT changing to `MethodHandle::changeReturnType`, I think doing that also begs for adding a `MethodHandle::changeParameterType` (with a single index and varargs overload). But, in that case it seems fair to just point users at `asType` instead.

In other words; `dropReturn` seems fine to get parity with `dropArguments`, but `changeReturnType` seems a step too close to `asType` (also since it implies adding `changeParameterType` as well). So, maybe this RFE is a bust, and users should use `asType` instead? Or do you think adding both `changeReturnType` and `changeParameterType`(s) seems worth it? (I'm not so sure).

**Comment by rose00** [comment]
> …`dropReturn` seemed like a good choice since we already have `dropArguments`. WRT changing to `MethodHandle::changeReturnType`...

That's a very reasonable point.  People might look for `dropRT` after they find `dropAs`.  And `MHs` is designed as a large-ish set of utility methods.

I agree that adding `MH::changeRT` is a slippery slope; it tends to lift more of the MT API onto MH, and it starts to put new stuff into the smaller MH API; that was my bad.

But (in the spirit of brainstorming, and agreeing with your present proposal), I'll suggest a separate idea to think about.  Use a HOFP API to give easy access to the entire `MT` API all in one go, by reifying the `MH.type` property as a temporary (lambda argument):

```
MethodHandle mapType(MethodHandle mh, UnaryOperator<MethodType> fn) {
   return mh.asType(fn.apply(mh.type()));
}
```

This also suggests a possible argument transform utility, a sort of mini-version of Charlie Nutter's MH builder API:

```
MethodHandle mapArguments(MethodHandle mh, UnaryOperator<List<ArgumentValue<?>>> fn) {
  var args = mh.type().argumentList().stream().map(t -> makeArg(t)).asList();
  args = fn.apply(args);
  … do stuff to re-weave mh with the resulting argument transforms …
}
public sealed interface ArgumentValue<T> {
  Class<T> type();
  <U> ArgumentValue<U> asType(Class<U> newType);
  … other stuff for pre-applying MHs … 
}
```

**Comment by forax** [comment]
instead of mapArguments(), i believe it's better to have a method map on a MethodType that takes two mapping functions, one for the return type and one for the prameter types.

so one can write:
  MHs.mapType(mh, mt -> mt.map(Function.identity(), t-> t.isPrimitive? t: Object.class))
to not change the return type and erase the parameter type to Object if they are not primitives

With that, changing the return type to void can be done that way too
  MHs.mapType(mh, mt -> mt.map(__ -> void.class, Function.identity())

**Comment by JornVernee** [comment]
It's an interesting idea. It turns something like:

```
target = target.asType(target.type().changeReturnType(void.class));
```

Into 

```
target = target.asType(mt -> mt.changeReturnType(void.class));
```

I.e. it removes the need to reference the target handle/type twice, which prevents them from potentially going out of sync (which seems more likely when you have multiple MethodTypes and MethodHandles flying around). Also, it feels unnecessary that, I have to re-specify `target`'s type as a base, when I'm asking `target` to change it's type. This also solves that.

**Comment by JornVernee** [comment]
CSR is now provisional. Please review: https://bugs.openjdk.java.net/browse/JDK-8255398

Thanks.
